concordium-consensus/src/Concordium/MultiVersion.hs

```diff
@@ -385,13 +539,59 @@ withWriteLock a = MVR $ \mvr -> withWriteLockIO mvr (runMVR a mvr)
-    bracketOnError (takeMVar mvWriteLock) (tryPutMVar mvWriteLock) $ \_ -> do
-        tid <- myThreadId
-        mvLog Runner LLTrace $ "Acquired global state lock on thread " ++ show tid
-        res <- a
-        putMVar mvWriteLock ()
-        mvLog Runner LLTrace $ "Released global state lock on thread " ++ show tid
-        return res
+    bracketOnError
+        (takeMVar mvWriteLock)
+        ( \() ->
+            tryPutMVar mvWriteLock ()
+                >> mvLog Runner LLError "Released global state lock following error."
+        )
+        $ \_ -> do
+            tid <- myThreadId
+            mvLog Runner LLTrace $ "Acquired global state lock on thread " ++ show tid
+            res <- a
+            putMVar mvWriteLock ()
+            mvLog Runner LLTrace $ "Released global state lock on thread " ++ show tid
+            return res
+
+-- |Perform an action while holding the global state write lock. Optionally, when the action
+-- completes, a thread is forked to perform a follow-up action before releasing the lock.
+-- If either the action or the follow-up action throws an exception, the write lock will be
+-- released.
+withWriteLockMaybeFork :: MVR finconf (a, Maybe b) -> (b -> MVR finconf ()) -> MVR finconf a
+withWriteLockMaybeFork action followup = MVR $ \mvr ->
+    withWriteLockMaybeForkIO mvr (runMVR action mvr) (\b -> runMVR (followup b) mvr)
+
+-- |Perform an action while holding the global state write lock. Optionally, when the action
+-- completes, a thread is forked to perform a follow-up action before releasing the lock.
+-- If either the action or the follow-up action throws an exception, the write lock will be
+-- released.
+withWriteLockMaybeForkIO :: MultiVersionRunner finconf -> IO (a, Maybe b) -> (b -> IO ()) -> IO a
+{-# INLINE withWriteLockMaybeForkIO #-}
+withWriteLockMaybeForkIO MultiVersionRunner{..} action followup = mask $ \unmask -> do
+    () <- takeMVar mvWriteLock
+    tid <- myThreadId
+    mvLog Runner LLTrace $ "Acquired global state lock on thread " ++ show tid
+    (res, mContinue) <- unmask action `onException` tryPutMVar mvWriteLock ()
+    case mContinue of
+        Just continueArg -> do
+            let release = do
+                    putMVar mvWriteLock ()
+                    tid' <- myThreadId
+                    mvLog Runner LLTrace $ "Released global state lock on thread " ++ show tid'
+            -- forkIO is guaranteed to be uninterruptible, so we can be sure that an async exception
+            -- won't prevent the lock being released. Also note that the masking state of the thread
+            -- is inherited, so we unmask when running the follow-up.
+            followupThread <-
+                forkIO (unmask (try @SomeException (followup continueArg)) >> release)
+            mvLog Runner LLTrace $
+                "Transferred global state lock from thread "
+                    ++ show tid
+                    ++ " to thread "
+                    ++ show followupThread
```

Comment:
The follow-up can involve finalizing blocks and potentially baking a block. Since this can be a fairly major operation, it is good to relay the message as soon as possible.